<h1> Комментарий ревьюера </h1>

Привет, Константин!
Меня зовут Виктория Ильина, и я буду проверять твой проект. Предлагаю общаться на "ты", но если тебе будет не комфортно, обязательно сообщи и мы перейдем на "вы". 

Моя основная цель - не указать на совершённые тобой ошибки, а поделиться своим опытом и помочь тебе стать специалистом по Data Science. Ты уже проделал большую работу над проектом, но давай сделаем его ещё лучше. Ниже ты найдёшь мои комментарии - пожалуйста, не перемещай, не изменяй и не удаляй их. Я буду использовать различные цвета, чтобы было легче воспринимать мои комментарии:

<div class="alert alert-block alert-danger">
<b>Комментарий ревьюера❌:</b> Так выделены самые важные замечания. Без их отработки проект не будет принят.
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюера🤔:</b> Так выделены небольшие замечания. Одно-два таких замечания в проекте допустимы, но если их будет больше - тебе будет необходимо внести исправления. Однако, если перед нами стоит цель не просто сдать проект и забыть, а приобрести новые знания и успешно применять их в будущем, я бы всё-таки рекомендовала тебе прорабатывать эти недочёты.
</div>
    
<div class="alert alert-block alert-success">
<b>Комментарий ревьюера👍:</b> Так я выделю все остальные комментарии.
</div>
    
Давай работать над проектом в диалоге: если ты что-то меняешь в проекте или отвечаешь на мои комментарии - пиши об этом. Мне будет легче отследить изменения, если ты выделишь свои комментарии:

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> Например, вот так.
</div>

Давай посмотрим, что получилось!

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span><ul class="toc-item"><li><span><a href="#Вывод-по-шагу-1" data-toc-modified-id="Вывод-по-шагу-1-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Вывод по шагу 1</a></span></li></ul></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span><ul class="toc-item"><li><span><a href="#Вывод." data-toc-modified-id="Вывод.-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Вывод.</a></span></li></ul></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span><ul class="toc-item"><li><span><a href="#Вывод." data-toc-modified-id="Вывод.-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Вывод.</a></span></li></ul></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span><ul class="toc-item"><li><span><a href="#Сгенерируем-матрицу-без-использования-функции,-проверяющей-обратимость-матрицы" data-toc-modified-id="Сгенерируем-матрицу-без-использования-функции,-проверяющей-обратимость-матрицы-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Сгенерируем матрицу без использования функции, проверяющей обратимость матрицы</a></span></li><li><span><a href="#Сгенерируем-матрицу-с-использованием-функции,-проверяющей-обратимость-матрицы" data-toc-modified-id="Сгенерируем-матрицу-с-использованием-функции,-проверяющей-обратимость-матрицы-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Сгенерируем матрицу с использованием функции, проверяющей обратимость матрицы</a></span></li><li><span><a href="#Создадим-класс-модели-линейной-регрессии-LinearRegression" data-toc-modified-id="Создадим-класс-модели-линейной-регрессии-LinearRegression-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Создадим класс модели линейной регрессии <em>LinearRegression</em></a></span></li><li><span><a href="#Используем-пайплайн-для-получения-метрики-R2" data-toc-modified-id="Используем-пайплайн-для-получения-метрики-R2-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Используем пайплайн для получения метрики <em>R2</em></a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Общий вывод</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

## Загрузка данных

In [1]:
import pandas as pd # для загрузки данных
import numpy as np # для операций с матрицамиъ
import warnings # для отключения предупреждений
warnings.simplefilter(action='ignore', category=UserWarning) # для отключения предупреждений
import seaborn as sns # для построения тепловой карты (хитмэп)
import matplotlib.pyplot as plt # для построения распределений переменных
from sklearn.model_selection import train_test_split # для разбиения данных на обучающую и тестовую выборки
from sklearn.metrics import r2_score # для расчёта метрики R2
from sklearn.pipeline import Pipeline # для построения пайплайн
from sklearn.base import BaseEstimator, TransformerMixin # для использования в пайплайн
from sklearn.preprocessing import FunctionTransformer # для использования в пайплайн
RANDOM_STATE = 12345

In [2]:
df = pd.read_csv('/datasets/insurance.csv')

In [3]:
df.head()

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


### Вывод по шагу 1
* Пропусков нет.
* Типы данных приведены.
* Используется верхний регистр.
* Лишних для модели данных нет.

    
<div class="alert alert-block alert-success">
<b>Комментарий ревьюера👍:</b> Хорошо, продолжаем
</div>

## Умножение матриц

В этом задании вы можете записывать формулы в *Jupyter Notebook.*

Чтобы записать формулу внутри текста, окружите её символами доллара \\$; если снаружи —  двойными символами \\$\\$. Эти формулы записываются на языке вёрстки *LaTeX.* 

Для примера мы записали формулы линейной регрессии. Можете их скопировать и отредактировать, чтобы решить задачу.

Работать в *LaTeX* необязательно.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

**Ответим на вопрос и обоснуем решение.**

   **Умножим признаки на обратимую матрицу. Изменится ли качество линейной регрессии? (Её можно обучить заново.)**

   **a. Изменится. Приведём примеры матриц.**

   **b. Не изменится. Укажем, как связаны параметры линейной регрессии в исходной задаче и в преобразованной.**

**Ответ:** не изменится.

**Обоснование:** для обоснования вывода формулы приведём основные свойства матриц.

**Основные свойства транспонирования матриц:**

1. Двойное транспонирование ***возвращает*** исходную матрицу: $(А^{T})^{T}=А$.

2. Транспонирование ***суммы*** матриц эквивалентно сумме транспонированных слагаемых: $(А+В)^{T}=А^{T}+В^{T}$.

3. Транспонирование ***произведения*** двух матриц эквивалентно произведению транспонированных матриц, взятых ***в обратном порядке***: $(АВ)^{T} = В^{T}А^{T}$.

4. Произведение матрицы на свою транспонированную: $А^{T}А$ или $АА^{T}$ всегда имеет результатом ***симметричную*** квадратную матрицу.

**Основные свойства обратной матрицы:**

1. $АА^{-1}=А^{-1}А=E$, где $E$ - единичная матрица.

2. $(АВ)^{-1}=В^{-1}А^{-1}$, т.е. при раскрытии скобок порядок сомножителей меняется на обратный.

3. $(А^{T})^{-1}=(А^{-1})^{T}$, т.е. операции обращения и транспонирования можно менять местами.

Умножение матриц ассоциативно:

$$
(ABC)^{-1} = C^{-1}(AB)^{-1} = (BC)^{-1}A^{-1} = C^{-1}B^{-1}A^{-1}
$$

НО! Умножение матриц некоммутативно. Это значит, что от перестановки местами матриц в произведении результат меняется.

Умножим матрицу признаков $X$ в формуле обучения $w(X) = (X^T X)^{-1} X^T y$ на обратимую матрицу $P$:

$$
w(XP) = [(XP)^T (XP)]^{-1} (XP)^T y = [P^T X^T XP]^{-1} P^T X^T y = P^{-1}[P^T (X^T X)]^{-1} P^T X^T y = P^{-1}(X^T X)^{-1} [(P^T)^{-1} P^T] X^T y = P^{-1}(X^T X)^{-1} X^T y,
$$

где $[(P^T)^{-1} P^T]=E$.


Отсюда укажем, как связаны параметры линейной регрессии в исходной задаче и в преобразованной:

$$
w(XP) = P^{-1} w(X)
$$

Подставим $w(XP)$ в формулу предсказания $a = Xw(X)$ (т.е. заменим $X$ на $XP$):

$$
a = XP w(XP) = XP [P^{-1} w(X)] = Xw(X)
$$

### Вывод. 

В результате выполнения задачи этого раздела было выявлено следующее:

мы доказали, что при умножении матрицы признаков $X$ на обратимую матрицу $P$ вектор весов $w$ меняется, а вектор целевого признака $a$ нет. Т.е. качество линейной регрессии не изменится.

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера👍:</b> Абсолютно верно, молодец, что не раскрыл $(X^T X)^{-1}$, так как для неквадратных матриц $X$ и $X^T$ обратных матриц не существует. - Это очень частая ошибка в этом проекте
</div>


## Алгоритм преобразования

**Предложим алгоритм преобразования данных для решения задачи. Обоснуем, почему качество линейной регрессии не поменяется.**

**Алгоритм**

1.	Разобьём исходные данные, выделив целевой признак.
2.	Разделим данные на обучающую и тестовую выборки.
3.	Сгенерируем случайную квадратную матрицу *P* с размером, равным количеству признаков в исходном датасете. Нужно генерировать матрицу до тех пор, пока мы не получим обратимую, и наш код гарантированно полностью не отработает. Добиться этого можно с помощью написания специальных для этого функций. Если сгенерированная матрица будет необратимой, то выскочит ошибка, код не исполнится до конца. Нужно генерировать матрицу до тех пор, пока не получим обратимую, и наш код гарантированно полностью не отработает. Добиться этого можно с помощью написания специальных для этого функций.
4.	Умножим признаки обучающей и тестовой  выборок  исходного датасета на сгенерированную случайную квадратную матрицу *P*.
5.	Создадим класс модели машинного обучения *LinearRegression*.
6.	Напишем функцию для создания модели линейной регрессии и получим её предсказания метрики *R2*.
7.	Получим *R2* модели для исходной и преобразованной матриц.

**Обоснование**

Обоснование было приведено в разделе 2 "Умножение матриц".

### Вывод. 

В результате выполнения задачи этого раздела был предложен алгоритм преобразования данных для решения задачи.

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера👍:</b> Верно
</div>


## Проверка алгоритма

**Запрограммируем этот алгоритм, применив матричные операции. Проверим, что качество линейной регрессии из *sklearn* не отличается до и после преобразования. Применим метрику *R2*.**

In [5]:
# Разобьём исходные данные, выделив целевой признак:
features = df.drop('Страховые выплаты', axis=1)
target = df['Страховые выплаты']

# Разделим данные на обучающую и тестовую выборки в соотношении 80:20:
features_train, features_test, target_train, target_test = train_test_split(
    features, target, test_size=0.2, random_state=RANDOM_STATE)

### Сгенерируем матрицу без использования функции, проверяющей обратимость матрицы

Сначала выполним задание без использования функции, проверяющей обратимость матрицы.
Квадратные матрицы, для которых можно найти обратные, называются обратимыми (англ. *invertible matrix*). 

Необратимые матрицы встречаются редко. Если сгенерировать случайную матрицу функцией *numpy.random.normal()* (англ. «нормальное распределение»), вероятность получить необратимую матрицу близка к нулю.

In [6]:
# Сгенерируем случайную квадратную матрицу P с размером, равным количеству признаков в исходном датасете,
# (т.к. количество столбцов исходной матрицы датасета должно совпадать с количеством строк матрицы P):
P = np.random.normal(size=(df.shape[1] - 1, df.shape[1] - 1))
display(P)

array([[ 1.94363868, -1.24322566,  1.93794196,  0.88725796],
       [ 1.2386124 , -0.56383335,  1.63994805,  0.06063893],
       [-0.23532751,  0.06985011, -1.06195639,  1.81215281],
       [-0.24032465, -0.82712136, -0.93569977,  0.52726864]])

Чтобы найти обратную матрицу, вызовем функцию *numpy.linalg.inv()* (от англ. *linear algebra*, «линейная алгебра»; *invert*, «обратить»). Также она поможет проверить матрицу на обратимость: если матрица необратима, будет обнаружена ошибка.

In [7]:
# Проверим, что эта матрица обратима (умножим её на обратную ей матрицу):
display(np.round(P @ np.linalg.inv(P)))

array([[ 1.,  0., -0.,  0.],
       [-0.,  1.,  0., -0.],
       [ 0., -0.,  1., -0.],
       [ 0., -0., -0.,  1.]])

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера👍:</b> Получена единичная матрица, всё в порядке =)
</div>

### Сгенерируем матрицу с использованием функции, проверяющей обратимость матрицы

In [8]:
# Вариант 1:

def get_rand_matrix():
    det = 0
    while det == 0:
        matrix = np.random.normal(size=(df.shape[1] - 1, df.shape[1] - 1))
        det = np.linalg.det(matrix)
    return matrix


mat = get_rand_matrix()

print(mat)

[[ 1.49726849 -0.22321926 -0.69877825 -0.18459511]
 [-1.51564661  0.62646139  0.6462998   0.28470039]
 [ 0.39357495  0.34972308 -0.62689246  0.49992683]
 [-0.05337907 -1.24745092 -0.20796027 -1.27435663]]


In [9]:
# Вариант 2 (с помощью рекурсии):

def func(dim):
    transform_matrix = np.random.rand(dim, dim)
    try: 
        inv = np.linalg.inv(transform_matrix)
        return transform_matrix, inv
    except:
        func(dim)
   

features_new, inv_P = func(df.shape[1] - 1)

print(inv_P)

[[-2.09201934  0.29246881  0.39894399  1.70986661]
 [-0.92505679 -1.08283139  0.39860254  1.92608929]
 [ 1.04042305  0.72848022  1.16475991 -2.22766874]
 [ 2.01954565  0.98090253 -0.79597915 -1.60244163]]


In [10]:
# Вариант 3 (с помощью рекурсии). 
# В этом варианте мы можем вывести и сгенерированную квадратную матрицу, 
# и преобразованные матрицы исходного датасета для обучающей и тестовой выборок:

def func(X, seed):
    try:
        state = np.random.RandomState(seed=seed)
        rand_matrix = state.normal(size=(X.shape[1], X.shape[1])) 
        inv = np.linalg.inv(rand_matrix)
        return X @ rand_matrix, inv
    except np.linalg.LinAlgError:
        func(X, seed)
        
        
features_train_P, inv_P = func(features_train, RANDOM_STATE)
features_test_P, inv_P = func(features_test, RANDOM_STATE)

display('Сгенерированная квадратная матрица P', inv_P)
display('Преобразованная исходная матрица для обучающей выборки', features_train_P.head())
display('Преобразованная исходная матрица для тестовой выборки', features_test_P.head())

'Сгенерированная квадратная матрица P'

array([[-1.31136747,  0.3921804 ,  0.18868055, -0.67088287],
       [ 1.75872714,  0.14106138, -0.17773045,  0.79787127],
       [-0.41702659, -0.22854768,  0.3550602 ,  0.33039819],
       [ 0.58912996,  0.19073027, -0.5545481 ,  0.6259302 ]])

'Преобразованная исходная матрица для обучающей выборки'

0             1             2             3
317   30646.963786  49636.428451  40090.700423 -51582.192361
4143  26378.998328  42684.283161  34449.072906 -44320.047816
4252  39154.876844  63382.295727  51170.233698 -65835.023067
710   30124.374233  48775.071382  39386.411136 -50672.583615
148   30047.196985  48650.198995  39284.339283 -50543.847934

'Преобразованная исходная матрица для тестовой выборки'

0             1             2             3
3183  30057.850862  48657.852763  39288.862573 -50539.779915
1071  33243.711676  53791.465256  43417.212539 -55851.269636
2640  32452.310831  52529.724173  42405.775923 -54560.476324
2282  26801.300964  43403.796943  35052.047801 -45102.859674
1595  30842.599674  49915.434746  40296.795194 -51833.747055

### Создадим класс модели линейной регрессии *LinearRegression*

In [11]:
class LinearRegression:
    def fit(self, train_features, train_target):
        X = np.concatenate((np.ones((train_features.shape[0], 1)), train_features), axis=1)
        y = train_target
        w = ((np.linalg.inv(X.T @ X)) @ X.T) @ y
        self.w = w[1:]
        self.w0 = w[0]

    def predict(self, test_features):
        return (test_features @ self.w) + self.w0

Здесь мы применили сокращённую запись формулы линейной регрессии: в обучающую выборку добавили единичный столбец (*numpy.ones()*).

In [12]:
# Напишем функцию для предсказания метрики R2, применив созданный выше класс модели линейной регрессии:
def model(x_train, y_train, x_test, y_test):
    model = LinearRegression()
    model.fit(x_train, y_train)
    predictions = model.predict(x_test)
    print('R2 модели:', r2_score(y_test, predictions))
    print(f'R2 модели (округлённо): {r2_score(y_test, predictions):.2f}')

In [13]:
# R2 модели для исходной матрицы:
model(features_train, target_train, features_test, target_test)

R2 модели: 0.41176839567704737
R2 модели (округлённо): 0.41


In [14]:
# R2 модели для преобразованной матрицы:
model(features_train_P, target_train, features_test_P, target_test)

R2 модели: 0.41176840632669975
R2 модели (округлённо): 0.41


### Используем пайплайн для получения метрики *R2*

Для получения метрики *R2* полезно использовать пайплайн.

В узком смысле, пайплайн – это модуль `sklearn.pipeline`, который позволяет автоматизировать предварительные преобразования данных перед обучением модели. 

Построим пайплайн с использованием созданного ранее класса модели линейной регрессии `LinearRegression`.

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера👍:</b> Здорово, что работаешь с пайплайнами, это очень поможет тебе в дальнейшем обучении и работе
</div>

In [15]:
# 1. Создаем класс MatrixTransformer:   
    
class MatrixTransformer(BaseEstimator, TransformerMixin):

    def __init__(self, variables):
        """В конструкторе сохраним имена колонок и сгененрируем матрицу преобразования"""
        self.variables = variables
        size = len(variables)
        rng = np.random.default_rng(seed=RANDOM_STATE)
        self.M = rng.normal(size=(size, size))

    def fit(self, X, y):
        return self

    def transform(self, X):
        X_tr = X.drop(self.variables, axis=1)
        X_tr[self.variables] = X[self.variables] @ self.M
        return X_tr

# 2. Иcпользуем Pipeline:

pipe = Pipeline([
    ('transformer', MatrixTransformer(features.columns)),
    ('model', LinearRegression()),
])

# Обучим модель линейной регрессии на исходных признаках датасета и получим её предсказания метрики R2:
pipe.fit(features_train, target_train)
predictions = pipe.predict(features_test)
r2_score(target_test, predictions)

0.4117684339148745

In [16]:
# Преобразуем датасет: умножим признаки обучающей и тестовой выборок исходного датасета 
# на сгенерированную случайную квадратную матрицу P:
features_train_P = features_train @ P
features_test_P = features_test @ P
display(features_train_P.head())
display(features_test_P.head())

0            1             2             3
317   -9341.743110  2767.103321 -42234.936767  72125.949051
4143  -7996.712545  2365.077076 -36251.372679  61978.938828
4252 -11897.196394  3520.942553 -53872.584715  92061.505895
710   -9167.104957  2713.697580 -41478.447709  70857.927197
148   -9143.331882  2707.539691 -41371.316370  70676.184648

0            1             2             3
3183  -9137.859794  2702.239160 -41365.923729  70678.069648
1071 -10081.165505  2980.693672 -45690.194413  78107.872483
2640  -9857.038449  2917.456749 -44642.468109  76294.885368
2282  -8164.624934  2419.507033 -36923.283416  63064.130476
1595  -9363.278400  2767.578600 -42414.760535  72490.707567

In [17]:
# После преобразования матриц названия признаков были изменены на индексы от 0 до 3.
# Восстановим эти названия для того, чтобы сработал наш пайплайн:

features_train_P.columns = ['Пол', 'Возраст', 'Зарплата', 'Члены семьи']
features_test_P.columns = ['Пол', 'Возраст', 'Зарплата', 'Члены семьи']

display(features_train_P.head())
display(features_test_P.head())

Пол      Возраст      Зарплата   Члены семьи
317   -9341.743110  2767.103321 -42234.936767  72125.949051
4143  -7996.712545  2365.077076 -36251.372679  61978.938828
4252 -11897.196394  3520.942553 -53872.584715  92061.505895
710   -9167.104957  2713.697580 -41478.447709  70857.927197
148   -9143.331882  2707.539691 -41371.316370  70676.184648

Пол      Возраст      Зарплата   Члены семьи
3183  -9137.859794  2702.239160 -41365.923729  70678.069648
1071 -10081.165505  2980.693672 -45690.194413  78107.872483
2640  -9857.038449  2917.456749 -44642.468109  76294.885368
2282  -8164.624934  2419.507033 -36923.283416  63064.130476
1595  -9363.278400  2767.578600 -42414.760535  72490.707567

In [18]:
# Обучим модель линейной регрессии на преобразованных признаках датасета и получим её предсказания метрики R2:
pipe.fit(features_train_P, target_train)
predictions = pipe.predict(features_test_P)
r2_score(target_test, predictions)

0.4117568334215683

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера👍:</b> Дополнительно можно попробовать декодировать данные обратно
</div>

## Общий вывод

Разработан такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию.

Входные данные - данные клиентов страховой компании «Хоть потоп».

Результаты исследования позволят защитить данные клиентов страховой компании «Хоть потоп».

Перед проведением исследования была сформулирована гипотеза:

- Гипотеза: при умножении признаков на обратимую матрицу качество линейной регрессии не меняется.

В результате удалось получить следующие результаты.

Гипотеза подтверждена:
- *0.41* - значение метрики *R2* модели для исходной матрицы;
- *0.41* - значение метрики *R2* модели для преобразованной матрицы.

<b>Общие рекомендации:</b>

Специалистам, которые занимаются защитой персональных данных клиентов страховой компании «Хоть потоп», можно применять разработанный метод преобразования данных (умножение матрицы признаков исходного датасета на случайную матрицу), т.к. при преобразовании качество моделей машинного обучения не ухудшается.

## Чек-лист проверки

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные загружены
- [x]  Выполнен шаг 2: получен ответ на вопрос об умножении матриц
    - [x]  Указан правильный вариант ответа
    - [x]  Вариант обоснован
- [x]  Выполнен шаг 3: предложен алгоритм преобразования
    - [x]  Алгоритм описан
    - [x]  Алгоритм обоснован
- [x]  Выполнен шаг 4: алгоритм проверен
    - [x]  Алгоритм реализован
    - [x]  Проведено сравнение качества моделей до и после преобразования

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера👍:</b> Спасибо тебе за проделанную работу, ты демонстрируешь отличное владение материалом спринта. Правда очень крепкая работа, достойная быть в портфолио. 
    
Отправляю обратно, чтобы ты мог задать какие-то вопросы, может хочешь какую-то подборку статей.   

Обзоры по применению криптографии в машинном обучении: https://arxiv.org/pdf/1902.04109.pdf, https://www.researchgate.net/publication/342746786_Some_Applications_of_Machine_Learning_in_Cryptography

Небольшой обзор на Шифр Хилла: https://habr.com/ru/post/595281/
    
Ещё пара статей на эту тему:
    
-	https://medium.com/intuit-engineering/machine-learning-on-encrypted-data-no-longer-a-fantasy-58e37e9f31d7
-	https://towardsdatascience.com/cryptography-crash-course-for-the-intimidated-892b7889528a
-	https://towardsdatascience.com/homomorphic-encryption-machine-learning-new-business-models-2ba6a4f185d

   
Примеры использования линейной алгебры в машинном обучении:
https://machinelearningmastery.com/examples-of-linear-algebra-in-machine-learning/
    
Для чего нужна математика в машинном обучении?
https://proglib.io/p/dlya-chego-nuzhen-matematicheskiy-analiz-v-mashinnom-obuchenii-2021-02-15

Большая статья по векторам:
https://neptune.ai/blog/understanding-vectors-from-a-machine-learning-perspective  

Можно ещё дополнительно посмотреть лекции от МФТИ https://www.youtube.com/playlist?list=PLk4h7dmY2eYHHTyfLyrl7HmP-H3mMAW08
    
Классная серия роликов по линейной алгебре на английском (можно включить автоматические русские субтитры): https://www.youtube.com/playlist?list=PLRDl2inPrWQW1QSWhBU0ki-jq_uElkh2a</div>


<div class="alert alert-block alert-success">
<b>Комментарий ревьюера V.2👍:</b> Удачи в будущих проектах и работе, всё обязательно получится!
</div>